activerecord/lib/active_record/future_result.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class FutureResult
+    Canceled = Class.new(ActiveRecordError)
+
+    delegate :empty?, to: :result!
+
+    def initialize(pool, *args, **kwargs)
+      @mutex = Mutex.new
+
+      @pool = pool
+      @args = args
+      @kwargs = kwargs
+
+      @executed = false
+      @error = nil
+      @result = nil
+      schedule!
+    end
+
+    def schedule!
+      ActiveRecord::Base.asynchronous_queries_tracker.register(self)
+      @pool.schedule_query(self)
+    end
+
+    def cancel!
+      @executed = true
+      @error = Canceled
+    end
+
+    def execute_or_skip
+      return if @executed
+
+      @pool.with_connection do |connection|
+        return unless @mutex.try_lock
+        begin
+          return if @executed
+          execute_query(connection, background: true)
+        ensure
+          @mutex.unlock
+        end
+      end
+    end
+
+    def result!
+      execute_or_wait
+      if @error
+        raise @error
+      else
+        @result
+      end
+    end
+
+    private
+      def execute_or_wait
+        return if @executed
+        if @mutex.try_lock
+          begin
+            execute_query(@pool.connection)
+          ensure
+            @mutex.unlock
+          end
+        else
+          @mutex.synchronize do
+            return if @executed
+            execute_query
```

Comment:
I made a mistake here. In theory this should never be reached, I should raise something instead.